src/java.base/share/classes/jdk/crac/RCULock.java

```diff
@@ -0,0 +1,324 @@
+package jdk.crac;
+
+import jdk.internal.ref.CleanerFactory;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.lang.invoke.*;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.TreeSet;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * This class represents a RW-lock with extremely lightweight read-locking
+ * and heavyweight write locking.
+ * <p>
+ * Readers should use the {@link #readLock()} and {@link #readUnlock()}; when
+ * the writer lock is not locked these are similar to calling no-op methods.
+ * The lock needs a declaration of all read-critical (top-level) methods;
+ * contrary to the regular synchronization pattern we need to
+ * <strong>acquire</strong> the read-lock <strong>inside</strong> the critical
+ * section (at the beginning) and <strong>release</strong> it <strong>outside</strong>,
+ * preferrably in the <code>finally</code> block.
```

Comment:
This assimetry in the interface does not look nice. Likely because the implementation, we can release the lock only outside a method with the critical code. But this leaks too much implementation details to the interface.

Reply:
It would *almost* work except for interpreted mode; in compiled code the safepoint poll is only at method exit (and in uncounted loops). I couldn't find the specific place in code but I've read that in interpreted mode the safepoint may happen on every bytecode instruction, so we would have to somehow disable this. Sounds too fragile to expect that you could jump straight into the critical section.
Alternatively we could try to do flow analysis on the bytecode and check where's the critical section, but that's also too much work just to make it look a tad better. This type of lock is intended for very specific use anyway.